src/smirnoffio.py

```diff
@@ -145,40 +146,93 @@ def assign_openff_parameter(ff, new_value, pid):
     if not hasattr(ff, '_forcebalance_assign_parameter_map'):
         ff._forcebalance_assign_parameter_map = dict()
     if pid not in ff._forcebalance_assign_parameter_map:
-        (handler_name, tag_name, value_name, smirks) = pid.split('/')
-        # Get the OpenFF parameter object
-        parameter = ff.get_parameter_handler(handler_name).parameters[smirks]
-        if hasattr(parameter, value_name):
+
+        if pid.startswith("/"):
+            # Handle the case were we are optimizing a handler attribute directly such
+            # as the 1-4 scaling factor.
+            handler_name, value_name = pid[1:].split('/')
+
+            # Get the OpenFF parameter handler.
+            parameter_container = ff.get_parameter_handler(handler_name)
+
+        else:
+            (handler_name, tag_name, value_name, smirks) = pid.split('/')
+
+            # Get the OpenFF parameter object
+            parameter_container = ff.get_parameter_handler(handler_name).parameters[smirks]
+
+        if hasattr(parameter_container, value_name):
             # If the value name is an attribute of the parameter then we set it directly.
-            unit = getattr(parameter, value_name).unit
             # Get the quantity of the parameter in the OpenFF forcefield object
-            param_quantity = getattr(parameter, value_name)
-        elif value_name in parameter._cosmetic_attribs:
+            param_quantity = getattr(parameter_container, value_name)
+
+        elif (hasattr(parameter_container, "_cosmetic_attribs") and
+              value_name in parameter_container._cosmetic_attribs):
+
+            param_quantity = None
+
+        else:
+            raise KeyError(
+                "The {} attribute is not supported by the {} handler".format(
+                    value_name, handler_name
+                )
+            )
+
+        # We can't use the caching approach when the parameter is a simple float,
+        # hence we set it directly here.
+        if isinstance(param_quantity, (float, int)):
+            setattr(parameter_container, value_name, new_value)
             param_quantity = None
         else:
-            # If the value name is a periodic attribute (say, k1) then we need to use
```

Comment:
This code should no longer be needed as the toolkit supports accessing `k1`, `k2` etc directly now.